tools/check_site.py

```diff
@@ -1,48 +1,164 @@
-                                    if not os.path.isdir(target) or not os.path.isfile(target + "/index.html"):
-                                        logger.error(f"Bad xref '{href}' in {ref_filename}")
-                except Exception as e:
-                    logger.error(f"Couldn't read HTML file {filename}")
-                    raise e
+import requests
+import threading
+
+all_local_files = {}
+external_links = {}
+
+
+# This runs in a separate thread (one per external URL)
+def is_valid_url(url: str, source_path: str) -> None:
+    """Check if an external URL is reachable."""
+    broken_source = source_path
+    try:
+        response = requests.head(url, allow_redirects=True, timeout=5)
+        if response.status_code < 400:
+            broken_source = ""
+    except requests.RequestException:
+            ...
+    external_links[url] = broken_source
+
+
+def check_links_in_html(file_path, file_dir, base_path, threads, stop_on_error: bool):
+    """Parse an HTML file and check all links."""
+    # Extract all href links from the file
+    links: list[str] = []
+    with open(file_path, "r", encoding="utf-8") as f:
+        content = f.read()
+        # Drop XML comments
+        content, _ = re.subn(r"<!--.*?-->", "", content, flags=re.DOTALL)
+        # Drop canonical link
+        content, _ = re.subn(r"<link\s+rel=\"canonical\"\s+href=\".*?\">\s*", "", content, flags=re.DOTALL)
+        # Find hrefs
+        refs = re.findall(r'href=(["\'])(.*?)\1', content)
+        if refs:
+            # Replace URL-encoded characters
+            refs = [r[1].replace("%20", " ").replace("%26", "&") for r in refs]
+            # Anchors in index.html
+            refs = [re.sub(r"^(.*/)#[^/]+$", r"\1index.html", r) for r in refs]
+            # Other anchors
+            refs = [re.sub(r"^(.*)#[^/]+$", r"\1", r) for r in refs]
+            # Links to directory (-> index.html) for local hrefs
+            refs = [re.sub(r"^((?!http).*/)$", r"\1index.html", r) for r in refs]
+            links.extend(refs)
+    broken_links = []
+
+    for href in links:
+        href = href.strip()
+        if href.startswith("/en"):
+            slash2 = href.find("/", href.find("/", 1) + 1)  # Find the third "/"
+            href = href[slash2 + 1 :]
+            if href == ".":
+                continue
+
+        if href.startswith("http"):  # External link
+            if (question_mark := href.find("?", 4)) != -1:
+                href = href[:question_mark]
+            if threads is not None and href not in external_links:
+                external_links[href] = ""
+                thread = threading.Thread(target=is_valid_url, args=(href, file_path))
+                thread.start()
+                threads.append(thread)
+
+        elif (
+            href.startswith("javascript:") or href.startswith("mailto:") or href.startswith("#")
+        ):  # Scrip, mail or internal ref link
+            ...
```

Comment:
pass ?

Reply:
Changed (and at another place as well). Thx